---
title: "ExAC Imperfects Analysis"
output: html_notebook
---

```{r setup, echo=F, message=F, warning = F}
library(dplyr)
library(tidyr)
library(ggplot2)
library(cowplot)
options(stringsAsFactors = F)

knitr::opts_chunk$set(warning = FALSE, message = F)
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 100), tidy = T)

setwd("~/Documents/projects/splicing/exac/notebooks")

data <- readRDS('../produced_data/exac_data_cons_exp.rds')
```


First, let's read in the read counts for the imperfect sequences and calculate an exon inclusion index.
```{r read_imperfect_counts}
imperfect_counts <- read.csv('../produced_data/imperfect_alignments.csv', header = T)
# only keep sort 2
imperfect_counts <- select(imperfect_counts, -ends_with('S1'))
imperfect_counts <- imperfect_counts %>% 
    separate(id, c('mutant_id', 'unknown', 'cigar_string', 'alignment'), sep = ' ') %>% 
    separate(mutant_id, c('ensembl_id', 'mutant_sub_id'), sep = '[.]', remove = F)
```

```{r normalize}
# normalize
imperfect_counts <- imperfect_counts %>%
    select(DP1S2:SP2S2) %>%
    # lower counts, reads per 100,000 instead of per million
    mutate_each(funs(norm = . / (sum(.) / 100000))) %>% 
    select(ends_with('norm')) %>% # get rid of original un-normalized columns
    bind_cols(select(imperfect_counts, mutant_id:alignment), .) %>% 
    rename(DP_R1 = DP1S2_norm, DP_R2 = DP2S2_norm, INT_R1 = INT1S2_norm, INT_R2 = INT2S2_norm,
           PS_R1 = R1.PS_norm, PS_R2 = R2.PS_norm, SP_R1 = SP1S2_norm, SP_R2 = SP2S2_norm) %>% 
    # filter low reads
    filter(DP_R1 >= 3 | DP_R2 >= 3 | INT_R1 >= 3 | INT_R2 >= 3 | PS_R1 >= 3 | PS_R2 >= 3 | SP_R1 >= 3 | SP_R2 >= 3)

# proportion of cells in each bin (relative to pre-sort)
# DP1S2, DP2S2, INT1S2, INT2S2, PS_R1, PS_R2, SP1S2, SP2S2
bin_prop <- c( 0.092, 0.091,  0.127, 0.122, 1, 1, 0.596, 0.603)

# multiply each bin count by bin proportion
imperfect_counts <- bind_cols(select(imperfect_counts, mutant_id:alignment), 
                 data.frame(mapply(`*`, select(imperfect_counts, DP_R1:SP_R2), bin_prop, SIMPLIFY = FALSE)))

```

```{r splicing_index}
imperfect_counts <- imperfect_counts %>%
    mutate(index_R1 = (DP_R1*0 + INT_R1*0.85 + SP_R1*1) / (DP_R1 + INT_R1 + SP_R1),
           index_R2 = (DP_R2*0 + INT_R2*0.85 + SP_R2*1) / (DP_R2 + INT_R2 + SP_R2))

# replace NaN with NA
imperfect_counts[imperfect_counts == 'NaN'] <- NA
```


```{r imperfect_index_viz}
imperfect_fit <- summary(lm(index_R2 ~ index_R1, imperfect_counts))$adj.r.squared 
ggplot(imperfect_counts, aes(index_R1, index_R2)) + geom_point(alpha = 0.50) +
    labs(x = 'exon inclusion index replicate 1', y = 'exon inclusion index replicate 2',
         title = 'Exon inclusion index for imperfect sequences') +
    annotate('text', label = paste('R^2==', round(imperfect_fit, 2)), parse = T, x = 0.95, y = 0.20)
```


```{r imperfect_index_hist_viz}
ggplot(imperfect_counts, aes(index_R1)) + geom_histogram(binwidth = 0.05) +
    labs(x = 'exon inclusion index replicate 1', 
         title = 'Distribution of imperfect exon inclusion index')
```

Let's use these imperfect counts to filter the imperfect reference file so we only score mutants that we have enough data for.

```{r}
imperfect_ref <- read.table('../ref/exac_imperfect_ref_sorted.txt', sep = '\t', header = F, col.names = c('header', 'sequence'))
imperfect_ref <- imperfect_ref %>% 
    separate(header, c('mutant_id', 'unknown', 'cigar_string', 'alignment'), sep = ' ')
imperfect_ref_filtered <- semi_join(imperfect_ref, imperfect_counts, by = 'mutant_id')
write.table(unite(imperfect_ref_filtered, header, mutant_id, unknown, cigar_string, alignment, sep = ' '),
            file = '../ref/exac_imperfect_ref_sorted_filtered.txt', sep = '\t', col.names = F, row.names = F, quote = F)
```


Let's join the imperfect data and the original data and see how their distributions compare.

```{r imperfect_vs_perfect_viz}
ggplot(bind_rows(select(data, id, index_R1) %>% mutate(type = 'perfect'),
                 mutate(imperfect_counts, type = 'imperfect') %>% 
                     select(id = mutant_id, index_R1, type)), aes(index_R1)) +
   geom_histogram(binwidth = 0.1, aes(fill = type), position = 'dodge') +
    scale_fill_manual(values = c('firebrick3', 'steelblue3'), name = '') +
    labs(x = 'exon inclusion index replicate 1')
```

```{r diff_psi_viz}
# calculate difference in splicing index between imperfect and perfect reference sequence
ggplot(imperfect_counts %>% 
    left_join(select(data, id, index_R1_perfect = index_R1), by = c('ensembl_id' = 'id')) %>% 
    mutate(dpsi = index_R1_perfect - index_R1), aes(dpsi)) +
    geom_histogram(binwidth = 0.10) + 
    labs(x = 'perfect PSI - imperfect PSI')
```

Let's get the imperfect data to look like our perfect data so we can incorporate it into future analysis. The ExAC library
consists of 1 nt SNP variants, so we will only consider imperfects that are different by 1 nt. We can find these variants by parsing the cigar string for those with only 1 mismatch (X - mismatch, I - insertion, D - deletion). For each imperfect variant, we need to add the matching intron/exon lengths from the corresponding perfect sequence, the imperfect sequence, and the location of the SNP (really an SNV). 

```{r format_imperfects}
cigar <- imperfect_counts$cigar_string
mismatches <- cigar[grepl('1X', cigar)]
mismatches <- mismatches[!grepl('D', mismatches)]

imperfect_snp <- imperfect_counts %>% 
    filter(stringr::str_count(cigar_string, '1X') == 1, 
           !grepl('D', cigar_string), 
           !grepl('I', cigar_string)) %>% 
    # add sequence
    left_join(select(imperfect_ref, mutant_id, sequence), by = 'mutant_id') %>% 
    left_join(select(data, id, perfect_sequence = sequence, chr:category, start_hg38, end_hg38), by = c('ensembl_id' = 'id')) %>% 
    # only keep those that appear in cleaned up data 
    filter(!is.na(start)) %>% 
    # only keep those whose sequence is 170bp
    filter(nchar(sequence) == 170)

count(imperfect_snp)


find_mismatch <- function(string1, string2) {
    a <- unlist(strsplit(string1, split = ''))
    b <- unlist(strsplit(string2, split = ''))
    
    loc <- which(a != b)
    
    if (length(loc) > 1) { return(-1) }
    else {return(loc) }
}

imperfect_snp$rel_position <- mapply(find_mismatch, imperfect_snp$sequence, imperfect_snp$perfect_sequence)

saveRDS(imperfect_snp, '../produced_data/exac_imperfects.rds', compress = T)
```

Let's take a quick look at the conservation of the "imperfect" SNP positions

```{r cons_write_input}
# add SNP position in hg38 coordinates
imperfect_snp <- imperfect_snp %>% 
    # start position always less than end position, regardless of strand
    mutate(snp_position_hg38 = start_hg38 + rel_position,
           snp_position_end_hg38 = snp_position_hg38 + 1) # need at least 1 bp interval

# output bed file of SNP positions 
write.table(file = '../produced_data/imperfect_snp_positions.bed', 
            x = imperfect_snp %>% 
                select(chr, snp_position_hg38, snp_position_end_hg38, mutant_id) %>% 
                na.omit() %>% 
                mutate(chr = paste0('chr', chr)), 
            sep = '\t', col.names = F, row.names = F, quote = F)
```

```{bash bedtools_sort}
bedtools sort -i ../produced_data/imperfect_snp_positions.bed > ../produced_data/imperfect_snp_positions_sorted.bed
```

```{python phast_cons, eval = F}
# usage: python phastCons_wrapper.py <sorted input bed file> <phastCons folder> <output file name> <num processes>
python ../scripts/phastCons_wrapper.py ../produced_data/imperfect_snp_positions_sorted.bed ../../ref/phastCons ../produced_data/imperfect_snp_positions_cons.bed
```

```{r conservation}
cons_scores <- read.table('../produced_data/imperfect_snp_positions_cons.bed', sep = '\t', 
                          col.names = c('chr', 'start', 'end', 'mutant_id', 'mean_cons_score'))

# convert NaN to NA
cons_scores[is.na(cons_scores)] <- NA

imperfect_snp <- left_join(imperfect_snp, select(cons_scores, mutant_id, mean_cons_score), by = 'mutant_id')

ggplot(imperfect_snp, aes(mean_cons_score)) + 
    geom_histogram() +
    labs(x = 'SNP conservation score', title = 'Conservation scores for imperfect variants')
```

Let's bring in the ExAC annotation

```{r exac_input}
# write file of regions of SNPs for input to tabix
write.table(imperfect_snp %>%
                mutate(chr = gsub('chr', '', chr),
                       snp_region = paste0(chr, ':', snp_position_hg38, '-', snp_position_end_hg38)) %>%
                select(snp_region) ,
            file = '../produced_data/tabix_input_imperfect_snp_regions.txt',
            quote = F, row.names = F, col.names = F, sep = '\t')
```

```{bash get_exac_annot, eval = F}
while read line; do
    tabix ../ref/ExAC.r0.3.1.sites.vep.vcf.gz $line >> ../ref/imperfect_snp_exac_annot.txt
done < ../produced_data/tabix_input_imperfect_snp_regions.txt
```

```{r exac}
exac_annot <- read.table('../ref/imperfect_snp_exac_annot.txt', sep = '\t', header = F)
colnames(exac_annot) <- c('chr', 'position', 'snp_id', 'ref_allele', 'alt_allele', 'quality', 'filter', 'info')
# get info column names
# split by ';', separates info columns
example_info <- unlist(strsplit(exac_annot$info[1], split = ';'))
# get names only, e.g. key=value
info_names <- unlist(lapply(strsplit(example_info, split = '='), `[[`, 1))
exac_annot <- separate(exac_annot, info, into = info_names, sep = ';') %>%
    # from info fields, only save CSQ, only one we care about right now
    select(chr:filter, CSQ)

# parse CSQ column, consequence annotations from Ensembl VEP (Variant Effect Predictor). Description of fields taken from header of vcf file
csq_fields <- 'Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON|INTRON|HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|Existing_variation|ALLELE_NUM|DISTANCE|STRAND|VARIANT_CLASS|MINIMISED|SYMBOL_SOURCE|HGNC_ID|CANONICAL|TSL|CCDS|ENSP|SWISSPROT|TREMBL|UNIPARC|SIFT|PolyPhen|DOMAINS|HGVS_OFFSET|GMAF|AFR_MAF|AMR_MAF|ASN_MAF|EAS_MAF|EUR_MAF|SAS_MAF|AA_MAF|EA_MAF|CLIN_SIG|SOMATIC|PHENO|PUBMED|MOTIF_NAME|MOTIF_POS|HIGH_INF_POS|MOTIF_SCORE_CHANGE|LoF_info|LoF_flags|LoF_filter|LoF|context|ancestral'
csq_fields <- unlist(strsplit(csq_fields, split = '[|]'))
exac_annot <- separate(exac_annot, CSQ, into = csq_fields, sep = '[|]') %>%
    mutate(Allele = gsub('CSQ=', '', Allele))

exac_annot <- exac_annot %>%
    mutate(annot = strsplit(Consequence, '&')) %>%
    unnest(annot) %>%
    select(chr:Allele, -Consequence, annot, IMPACT:ancestral) %>%
    distinct()

# join exac data
imperfect_snp <- imperfect_snp %>%
        mutate(chr = as.numeric(gsub('chr', '', chr))) %>% 
        left_join(select(exac_annot, chr, position, annot),
                 by = c('chr', 'snp_position_hg38' = 'position'))

count(exac_annot)
table(imperfect_snp$annot)
```

Only 9 imperfect variants have ExAC annotation and only 1 has functional annotation as an intron variant.

